### DISCUSSION TITLE
Lower latency

### DISCUSSION THREAD

**Discussion by fgaafgb** [ExpressLRS/ExpressLRS#441] [Ideas]
Is it possible to send channels that are changed _more_ frequently and channels that are unchanged _less_ frequently?

For example:

- channel present bitmask <4 bits>
- 1st channel value <11 bits> if present in bitmask
- 2st channel value <11 bits> if present in bitmask
- 3st channel value <11 bits> if present in bitmask
- 4st channel value <11 bits> if present in bitmask

So if you change the throttle only:

- 1000 if throttle is on the first channel
- 00000000000 if throttle value is at zero

So this would result in (if the normal packet is 8 bytes):

- 1 channel changing, 2 bytes + 2 bytes overhead - reduced latency with 50%
- 2 channel changing, 4 bytes + 2 bytes overhead - reduced latency with 25%
- 3 channel changing, 5 bytes + 2 bytes overhead - reduced latency with 12.5%
- 4 channel changing, 6 bytes + 2 bytes overhead - same

The latency isn't constant anyway since the value it sampled at descrete time steps and the reaction is not. There also need to be a maximum latency for constant channels, like 25 Hz or something. It could be implemented using round robin and only send channels that are changed OR that wasn't sent for 40 ms.

This example does not account for one bit switches so maybe there need to be 5 channels or a packet type.

**Comment by greenbigfrog** [comment]
on a quad in a race/FS scenario, you'll be changing all 4 channels that are sent with every packet (close to) constantly.

Perhaps in LR this could allow for higher rate, but current testing is showing that even if higher rates one can get quite enough range.

**Comment by JBKingdon** [comment]
It's an interesting idea, and is a little like what happens with hybrid switches where if the switches are unchanged they are sent round robin, but when a switch is changed it jumps the queue.

Applying it to the primary channels has a couple of catches. One is that they change a lot in normal flight, and that will become even more true as RC resolution is increased. The second is that ELRS is optimised around a fixed packet size, so we'd only be able to benefit from not sending a channel by first moving to variable packets, which would make the 'normal' (or at least the worst) case be slower. Since our current packets only just fit into the transmission windows we'd have to make compromises on either packet rate or range to accommodate that worst case.
